src/util/obbt.jl

```diff
@@ -0,0 +1,406 @@
+
+    return modifications
+
+end 
+
+function create_modifications(pm::GenericPowerModel, lb::Float64, ub::Float64, id::Tuple{Int64,Int64})
+
+    modifications = Dict{String,Any}()
+
+    modifications["per_unit"] = true
+    modifications["branch"] = Dict{String,Any}()
+
+    branches = ids(pm, :branch)
+
+
+    for branch in branches 
+        index = string(ref(pm, :branch, branch, "index"))
+        f_bus = ref(pm, :branch, branch, "f_bus") 
+        t_bus = ref(pm, :branch, branch, "t_bus")
+        bp = (f_bus, t_bus)
+        if bp == id 
+            modifications["branch"][index] = Dict{String,Any}( "angmin" => lb, "angmax" => ub )
+        end
+    end 
+
+    return modifications
+
+end
+
+function run_obbt_opf(file::String, model_constructor, solver; kwargs...)
+    data = PowerModels.parse_file(file)
+    return run_obbt_opf(data, model_constructor, solver; kwargs...)
+end 
+
+function run_obbt_opf(data::Dict{String,Any}, model_constructor, solver;
+    max_iter = 100, 
+    time_limit = 3600.0,
+    upper_bound = Inf,
+    upper_bound_constraint = false, 
+    sequential_obbt = false,
+    rel_gap_tol = Inf,
+    min_bound_width = 1e-2,
+    improvement_tol = 1e-3, 
+    output_tol = 1e+5,
+    termination = "avg",
+    kwargs...)
+
+    info(LOGGER, "maximum OBBT iterations set to default value of $max_iter")
+    info(LOGGER, "maximum time limit for OBBT set to default value of $time_limit seconds")
+
+    model_relaxation = build_generic_model(data, model_constructor, PowerModels.post_opf)
+    (ismultinetwork(model_relaxation)) && (error(LOGGER, "OBBT is not supported for multi-networks"))
+    (ismulticonductor(model_relaxation)) && (error(LOGGER, "OBBT is not supported for multi-phase networks"))
+
+    # check for model_constructor compatability with OBBT
+    check_variables(model_relaxation)
+
+    # check for other keyword argument consistencies 
+    check_obbt_options(upper_bound, rel_gap_tol, upper_bound_constraint)
+
+    # check termination norm criteria for obbt 
+    (termination != "avg" && termination != "max") && (error(LOGGER, "OBBT termination criteria can only be max or avg"))
+
+    # pass status 
+    status_pass = [:LocalOptimal, :Optimal]
+
+    # compute initial relative gap between relaxation objective and upper_bound 
+    result_relaxation = solve_generic_model(model_relaxation, solver)
+    current_relaxation_objective = result_relaxation["objective"]
+    if upper_bound < current_relaxation_objective
+        error(LOGGER, "the upper bound provided to OBBT is not a valid ACOPF upper bound")
+    end 
+    if !(result_relaxation["status"] in status_pass) 
+        error(LOGGER, "relaxation is infeasible, implying the ACOPF is infeasible")
+    end 
+    current_rel_gap = Inf 
+    (!isinf(rel_gap_tol)) && (current_rel_gap = (upper_bound - current_relaxation_objective)/upper_bound)
+    info(LOGGER, "Initial relaxation gap = $current_rel_gap")
```

Comment:
Maybe only print this if `upper_bound` is specified.

Reply:
done